### PR TITLE
Feature/firebase

### PR DESCRIPTION
**Pull request by paweljanicki**
## What was done
- Initialise Firebase project
- Create Firebase Playground
- Create FirebaseService
- Handle authentication to Firebase
- Add Firebase function to create custom token for authentication
- Add npm scripts to work with firebase
- Deploy Firebase Github Actions workflow
  - Deploys Firebase functions and Firestore rules to the environment specified by branch (to be triggered by push to development, staging and master branches)

## Testing
Instructions for the firebase setup in the README, also available on Notion https://www.notion.so/primedao/Firebase-setup-6753bec706d044cebf279e5ef3b5f9a8#0fe9abaec73648baa65b06b1a31cdabe

### CODE REVIEW

File: .github/workflows/firebase.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+name: Deploy Firebase
+on:
+  push:
+    branches: [ 'feature/firebase' ]
```
Comment (paweljanicki) [inline .github/workflows/firebase.yml:4]: Ultimately this should changed, very likely to `branches: [ development, staging, master ]` To be discussed.

File: package.json [hunk 2]
```diff
@@ -25,7 +25,15 @@
     "e2e-run": "npx cypress run",
     "lint": "eslint --ext .ts ./src ./test",
     "lint.fix": "npm run lint -- --fix",
-    "fetchContracts": "node scripts/fetchContracts.js"
+    "fetchContracts": "node scripts/fetchContracts.js",
+    "firebase": "npm run firebase-functions:watch & npm run firebase-emulators:start",
+    "firebase-export": "firebase emulators:export --force firebase-seed",
+    "firebase-emulators:start": "firebase emulators:start --project development --import=firebase-seed",
+    "firebase-functions:watch": "cd functions && npm run build:watch",
+    "firebase-deploy:feature/firebase": "firebase deploy --project=development",
```
Comment (paweljanicki) [inline package.json:33]: this script should eventually be deleted

File: functions/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import * as functions from "firebase-functions";
+import * as firebaseAdmin from "firebase-admin";
+import * as corsLib from "cors";
+import { getAddress } from "ethers/lib/utils";
+
+const admin = firebaseAdmin.initializeApp();
+
+const cors = corsLib({
```
Comment (dkent600) [inline functions/src/index.ts:8]: can we add a comment on why this is needed?  Does it affect the entire app?

File: functions/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  origin: true,
+});
+
+export const createCustomToken = functions.https.onRequest(
```
Comment (dkent600) [inline functions/src/index.ts:12]: can we comment in the code what a "custom token" is for?

File: tsconfig.json [hunk 5]
```diff
@@ -22,7 +22,8 @@
   },
   "include": [
     "src",
-    "test"
+    "test",
+    "functions/src"
```
Comment (dkent600) [inline tsconfig.json:26]: can we name the "functions" folder "firebaseFunctions" or "databaseFunctions" or "serverlessFunctions", something more specific?

**Review by dkent600 — CHANGES_REQUESTED** [review]

**Comment by dkent600** [comment]
Would be good to move the Notion information into a readme in the repo

**Comment by dkent600** [comment]
Are there any automated tests?

File: functions/src/index.ts (hunk 3 above)
Comment (paweljanicki) [reply functions/src/index.ts]: Commented in the code. It only affects the functions that use it, and it's necessary to make HTTP calls to a function

File: functions/src/index.ts (hunk 4 above)
Comment (paweljanicki) [reply functions/src/index.ts]: commented

File: tsconfig.json (hunk 5 above)
Comment (paweljanicki) [reply tsconfig.json]: renamed

**Review by dkent600 — APPROVED** [review]

**Pull request merged**